udp/udp.c

```diff
@@ -0,0 +1,243 @@
+// Copyright 2016 by BeeGee, see LICENSE.txt
+//
+// Adapted from: github.com/tuanpmt/esp_bridge, Created on: Mar 4, 2015, Author: Minh
+// Adapted from: rest.c, Author: Thorsten von Eicken
+
+#include "esp8266.h"
+#include "c_types.h"
+#include "ip_addr.h"
+#include "udp.h"
+#include "cmd.h"
+
+#define UDP_DBG
+
+#ifdef UDP_DBG
+#define DBG_UDP(format, ...) os_printf(format, ## __VA_ARGS__)
+#else
+#define DBG_UDP(format, ...) do { } while(0)
+#endif
+
+typedef struct {
+	char			*host;
+	uint32_t		port;
+	ip_addr_t		ip;
+	struct espconn	*pCon;
+	char			*data;
+	uint16_t		data_len;
+	uint16_t		data_sent;
+	uint32_t		resp_cb;
+	uint8_t			conn_num;
+} UdpClient;
+
+
+// Connection pool for UDP socket clients. Attached MCU's just call UDP_setup and this allocates
+// a connection, They never call any 'free' and given that the attached MCU could restart at
+// any time, we cannot really rely on the attached MCU to call 'free' ever, so better do without.
+// Instead, we allocate a fixed pool of connections an round-robin. What this means is that the
+// attached MCU should really use at most as many UDP connections as there are slots in the pool.
+#define MAX_UDP 4
+static UdpClient udpClient[MAX_UDP];
+static uint8_t udpNum = 0xff; // index into udpClient for next slot to allocate
```

Comment:
Cool. I thinks specially if we add flow-control and buffering it would be good to have these combined. Maybe the rest stuff can even be folded in as well.